### PR TITLE
Add support for ESM plugins

### LINKED ISSUE #7327: Support ESM plugins
### What is the problem you're trying to solve?

Currently, we support a plugin module only for CommonJS (CJS). It would be useful if we could support ESM as well.

https://github.com/stylelint/stylelint/blob/7cd8d00705e81cbe69bdbac85fd038cb26290e20/lib/augmentConfig.mjs#L323

### What solution would you like to see?

In addition to CJS, we should be able to import a plugin written in ESM. Probably, dynamic import (`import()`) may help implement it.

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Closes #7327

> Is there anything in the PR that needs further explanation?

In addition, this change does the following:
- improve the migration guide for v16
- rewrite the writing plugin guide for ESM
- make the changelog items more consistent

Note: I realized that #7279 produced a breaking change in `stylelint.utils.checkAgainstRule`.

### CODE REVIEW

File: .changeset/twelve-spies-design.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": major
+---
+
+Changed: `stylelint.utils.checkAgainstRule` to become an async function
```
Comment (jeddy3) [inline .changeset/twelve-spies-design.md:5]:
```suggestion
Changed: `stylelint.utils.checkAgainstRule` to be an async function
```

File: docs/developer-guide/plugins.md [hunk 2]
```diff
@@ -14,36 +14,47 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
 
 ## The anatomy of a plugin
 
+Let's say we write a plugin forbidding certain word (e.g., "foo") within a selector.
+The plugin's implementation would be:
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:18]: This example plugin disallows the word "foo" in selectors:

File: docs/developer-guide/plugins.md [hunk 3]
```diff
@@ -52,7 +63,31 @@ ruleFunction.ruleName = ruleName;
 ruleFunction.messages = messages;
 ruleFunction.meta = meta;
 
-module.exports = stylelint.createPlugin(ruleName, ruleFunction);
+export default stylelint.createPlugin(ruleName, ruleFunction);
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:66]:
```suggestion
export default createPlugin(ruleName, ruleFunction);
```

File: docs/developer-guide/plugins.md [hunk 4]
```diff
@@ -14,36 +14,47 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
...
+const ruleFunction = (primary, secondaryOptions) => {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, {
+      actual: primary,
+      possible: [true]
+    });
 
     if (!validOptions) {
       return;
     }
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:44]:
```suggestion
    if (!validOptions) return;
```

To match our code style.

File: docs/developer-guide/plugins.md [hunk 5]
```diff
@@ -14,36 +14,47 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
...
+
+      if (selector.includes("foo")) {
+        report({
+          result,
+          ruleName,
+          message: messages.rejected(selector),
+          node: ruleNode
+          word: selector,
+        });
+      }
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:57]:
```suggestion
      if (!selector.includes("foo")) return;
      
      report({
        result,
        ruleName,
        message: messages.rejected(selector),
        node: ruleNode
        word: selector,
      });
```

minor to advocate for our pattern of returning early in rules.

File: docs/developer-guide/plugins.md [hunk 6]
```diff
@@ -52,7 +63,31 @@ ruleFunction.ruleName = ruleName;
 ruleFunction.messages = messages;
 ruleFunction.meta = meta;
 
-module.exports = stylelint.createPlugin(ruleName, ruleFunction);
+export default stylelint.createPlugin(ruleName, ruleFunction);
+```
+
+> [!IMPORTANT]
+> A CommonJS plugin also works, but such a plugin may be unsupported in the future.
+> We recommend writing an ES module plugin if compatibility doesn't matter.
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:71]:
```suggestion
```

Let's take this out for now, and add something about commonjs plugins being deprecated in a follow-up issue.

File: docs/developer-guide/plugins.md [hunk 7]
```diff
@@ -75,52 +110,21 @@ You'll have to [learn about the PostCSS API](https://api.postcss.org/).
...
-  return (postcssRoot, postcssResult) => {
-    const validOptions = stylelint.utils.validateOptions(
-      postcssResult,
-      ruleName,
-      {
-        /* .. */
-      }
-    );
+    // load a forbidden word asynchronously
+    const forbiddenWord = await import("./forbidden-word.js");
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:121]:
```suggestion
    // load disallowed words asynchronously
    const disallowedWords = await import("./disallowed-words.js");
```

our consistent language

File: docs/developer-guide/plugins.md [hunk 8]
```diff
@@ -75,52 +110,21 @@ You'll have to [learn about the PostCSS API](https://api.postcss.org/).
...
-      // some async operation
-      setTimeout(() => {
-        // ... some logic ...
-        stylelint.utils.report({
-          /* .. */
-        });
-        resolve();
-      }, 1);
-    });
+    // report a warning if a violation against the word is detected...
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:125]:
```suggestion
    // report a warning if a problem word is detected...
```

File: docs/developer-guide/plugins.md [hunk 9]
```diff
@@ -245,49 +251,55 @@ It accepts an options object and a callback that is invoked with warnings from t
...
+};
 ```
 
 ## `stylelint.rules`
 
-All of the rule functions are available at `stylelint.rules`. This allows you to build on top of existing rules for your particular needs.
+All of the rule functions are available at the `stylelint.rules` object. This allows you to build on top of existing rules for your particular needs.
+
+> [!NOTE]
+> Every value in the `stylelint.rules` object is a `Promise` resolving a rule function. This avoids loading all the rules and slowing the start-up down.
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:302]:
```suggestion
> Every value in the `stylelint.rules` object is a `Promise` resolving a rule function.
```

removed internal implementation rationale.

File: docs/developer-guide/plugins.md [hunk 10]
```diff
@@ -296,17 +308,16 @@ All rules share a common signature. They are a function that accepts two argumen
 Here's an example of a plugin that runs `declaration-no-important` only if there is a special directive `@@check-declaration-no-important` somewhere in the stylesheet:
 
 ```js
-module.exports = stylelint.createPlugin(ruleName, (expectation) => {
-  const runDeclarationNoImportant = stylelint.rules[
-    "declaration-no-important"
-  ].then((rule) => rule(expectation));
+stylelint.createPlugin(ruleName, (primary) => {
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:311]:
```suggestion
createPlugin(ruleName, (primary) => {
```

File: docs/developer-guide/plugins.md [hunk 11]
```diff
@@ -245,49 +251,55 @@ It accepts an options object and a callback that is invoked with warnings from t
...
 - `context?`: the [context](rules.md#add-autofix) for the rule you are invoking
 
-Use the warning to create a _new_ warning _from your plugin rule_ that you report with `stylelint.utils.report`.
+Use the warning to create a _new_ warning _from your plugin rule_ that you report with [`stylelint.utils.report`](#stylelintutilsreport).
 
-For example, imagine you want to create a plugin that runs `at-rule-no-unknown` with a built-in list of exceptions for at-rules provided by your preprocessor-of-choice:
+For example, imagine you want to create a plugin that runs [`at-rule-no-unknown`](../../lib/rules/at-rule-no-unknown/README.md) with a built-in list of exceptions for at-rules provided by your preprocessor-of-choice:
 
 ```js
+const { checkAgainstRule, report } = stylelint.utils;
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:259]:
```suggestion
const { utils: checkAgainstRule, report } = stylelint;
```

File: lib/__tests__/fixtures/plugin-selector-no-foo.mjs [hunk 12]
```diff
@@ -0,0 +1,40 @@
+import stylelint from '../../index.mjs';
+
+const { report, ruleMessages, validateOptions } = stylelint.utils;
```
Comment (jeddy3) [inline lib/__tests__/fixtures/plugin-selector-no-foo.mjs:3]:
```suggestion
const {
  createPlugin,
  utils: { report, ruleMessages, validateOptions },
} = stylelint;
```

File: lib/__tests__/fixtures/plugin-selector-no-foo.mjs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+				});
+			}
+		});
+	};
+};
+
+rule.ruleName = ruleName;
+rule.messages = messages;
+
+export default stylelint.createPlugin(ruleName, rule);
```
Comment (jeddy3) [inline lib/__tests__/fixtures/plugin-selector-no-foo.mjs:40]:
```suggestion
export default createPlugin(ruleName, rule);
```

File: lib/__tests__/fixtures/plugin-selector-no-foo.mjs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [true],
+		});
+
+		if (!validOptions) {
+			return;
+		}
```
Comment (jeddy3) [inline lib/__tests__/fixtures/plugin-selector-no-foo.mjs:20]:
```suggestion
		if (!validOptions) return;
```

File: lib/__tests__/fixtures/plugin-selector-no-foo.mjs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+			const { selector } = ruleNode;
+
+			if (selector.includes('foo')) {
+				report({
+					result,
+					ruleName,
+					message: messages.rejected(selector),
+					node: ruleNode,
+				});
+			}
```
Comment (jeddy3) [inline lib/__tests__/fixtures/plugin-selector-no-foo.mjs:32]:
```suggestion
			if (!selector.includes('foo')) return;
			report({
				result,
				ruleName,
				message: messages.rejected(selector),
				node: ruleNode,
			});
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Fantastic! @ybiquitous It's incredible to see this, and it was only made possible by your immense effort to migrate the entire codebase to ESM!

I've tested this branch locally with a mix of ESM and CommonJS plugins. It all worked well.

I've made some minor suggestions to the wording and code style. 

Let's also add something to the migration guide. We can adopt the [`v15` format](https://github.com/stylelint/stylelint/blob/main/docs/migration-guide/to-15.md) of "significant or breaking changes". For example:

```markdown
# Migrating to 16.0.0

This release contains significant or breaking changes.

## Significant changes

- added support for ESM plugins

### Added support for ESM plugins

You can now create ESM plugins for Stylelint. This involved a year-long effort to migrate our codebase to ESM. We've updated our [developer guide](../developer-guide/plugins.md) with examples of ESM syntax.

## Breaking changes

This release contains breaking changes. We've:

- removed deprecated stylistic rules
- removed support for Node.js less than 18.12.0
- changed `.js` extension to `.mjs` and `.cjs`
- changed Node.js API returned resolved object
- changed Node.js API `stylelint.formatters` object
- changed Node.js API `stylelint.rules` object
- changed Node.js API `stylelint.utils.checkAgainstRule` function
- changed CLI to print problems to stderr
- changed CLI exit code for flag errors

### Removed deprecated stylistic rules

...
```

The rest of the headings will need to be bumped down a level too.

Let's do another prerelease after this PR is merged.

I'll create a follow-up issue about deprecating CommonJs plugins.

File: docs/developer-guide/plugins.md (hunk 2 above)
Comment (ybiquitous) [reply docs/developer-guide/plugins.md]: Thanks. Fixed via 3288290af.

File: docs/developer-guide/plugins.md [hunk 16]
```diff
@@ -14,36 +14,47 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
...
 
+Let's say we write a plugin forbidding certain word (e.g., "foo") within a selector.
+The plugin's implementation would be:
+
 ```js
-// Abbreviated example
-const stylelint = require("stylelint");
+import stylelint from "stylelint";
+
+const { report, ruleMessages, validateOptions } = stylelint.utils;
```
[the remark this answers is not in the document]
Comment (ybiquitous) [reply docs/developer-guide/plugins.md]: Agree. Fixed via 06b94d8b9.

File: docs/developer-guide/plugins.md (hunk 5 above)
Comment (ybiquitous) [reply docs/developer-guide/plugins.md]: Fixed bde97188d.

File: lib/__tests__/fixtures/plugin-selector-no-foo.mjs (hunk 12 above)
Comment (ybiquitous) [reply lib/__tests__/fixtures/plugin-selector-no-foo.mjs]: Fixed via 3ad3726ef.

**Comment by ybiquitous** [comment]
@jeddy3 Thanks a lot for the review. I've addressed and resolved all your review comments. 👌🏼 

In addition, I've pushed one more improvement (101f65e89) in the docs to clarify the `stylelint.utils` function notations, appending `()` like this:

```diff js
-`stylelint.utils.report`
+`stylelint.utils.report()`
```

Any thoughts?

**Comment by jeddy3** [comment]
The changes look good. As part of this PR, let's [update the migration guide](https://github.com/stylelint/stylelint/pull/7339#pullrequestreview-1731575003).

File: docs/developer-guide/plugins.md [hunk 17]
```diff
@@ -14,36 +14,47 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
...
 
-    // ... some logic ...
-    stylelint.utils.report({
-      /* .. */
+      report({
+        result,
+        ruleName,
+        message: messages.rejected(selector),
+        node: ruleNode,
+        word: selector
```
Comment (Mouvedia) [inline docs/developer-guide/plugins.md:56]: question

Should `index` and `endIndex` be mentioned somewhere?

File: docs/developer-guide/plugins.md [hunk 18]
```diff
@@ -143,100 +142,103 @@ testRule({
...
 
 Validates the options for your rule.
 
-### `stylelint.utils.checkAgainstRule`
+### `stylelint.utils.checkAgainstRule()`
 
 Checks CSS against a standard or custom Stylelint rule _within your own rule_. This function provides power and flexibility for plugins authors who wish to modify, constrain, or extend the functionality of existing Stylelint rules.
 
+> [!NOTE]
+> This is an async function. Your custom rule may need to wait until a `Promise` the function returns is resolved.
```
Comment (Mouvedia) [inline docs/developer-guide/plugins.md:240]: > a `Promise` the function returns

There's a typo here.

**Comment by ybiquitous** [comment]
> As part of this PR, let's [update the migration guide](https://github.com/stylelint/stylelint/pull/7339#pullrequestreview-1731575003).

Done! acd12edc1

File: docs/developer-guide/plugins.md (hunk 17 above)
Comment (ybiquitous) [reply docs/developer-guide/plugins.md:56]: Good catch. I've tried adding a new description to the `report()` sections (f83157c92). Any thoughts?

File: docs/developer-guide/plugins.md (hunk 18 above)
Comment (ybiquitous) [reply docs/developer-guide/plugins.md]: Good catch. Fixed via b703a8f0e.

File: docs/developer-guide/plugins.md (hunk 17 above)
Comment (Mouvedia) [reply docs/developer-guide/plugins.md:56]: > Any thoughts?

> For example:

has no examples.

> _Do not use…

Maybe use an `[!IMPORTANT]` instead?

Comment (ybiquitous) [reply docs/developer-guide/plugins.md:56]: Thanks. Fixed the typos and changed to `[!IMPORTANT]` (af6d74393)

**Review by jeddy3 — APPROVED** [review]
Wonderful, LGTM.

(I think we can merge and do another prerelease.)

**Comment by ybiquitous** [comment]
Okay, I'll merge this PR and release `16.0.0-2` from now.

**Pull request merged**